src/atomic.rs

```diff
@@ -0,0 +1,166 @@
+use std::marker::PhantomData;
+use std::mem;
+use std::sync::atomic::{AtomicPtr, Ordering};
+use std::sync::Arc;
+
+use crate::collector::{GcData, InternalGcRef, COLLECTOR};
+use crate::{Gc, Scan};
+
+/// An atomic `Gc<T>`, useful for concurrent algorithms
+///
+/// This has more overhead than an `AtomicPtr`, but cleanly handles memory management. It also is
+/// similar to `Gc<T>` in that it can be cloned, and therefore easily shared.
+///
+/// `AtomicGc` should be fairly fast, but you may not assume it does not block. In fact in the
+/// presence of an active garbage collection operation, all operations will block.
+#[derive(Clone, Debug)]
+pub struct AtomicGc<T: Scan> {
+    atomic_ptr: Arc<AtomicPtr<GcData>>,
+    backing_handle: InternalGcRef,
+    _mark: PhantomData<Gc<T>>,
+}
+
+impl<T: Scan> AtomicGc<T> {
+    pub fn new(v: &Gc<T>) -> Self {
+        let _collection_blocker = COLLECTOR.get_atomic_guard_spinlock_inclusive();
+
+        let data = v.internal_handle_ref().data();
+        let data_ptr = Arc::into_raw(data);
+        let atomic_ptr = Arc::new(AtomicPtr::new(data_ptr as _));
+
+        Self {
+            atomic_ptr: atomic_ptr.clone(),
+            backing_handle: COLLECTOR.new_handle_for_atomic(atomic_ptr),
+            _mark: PhantomData,
+        }
+    }
+
+    pub(crate) fn internal_handle(&self) -> InternalGcRef {
+        self.backing_handle.clone()
+    }
+
+    #[must_use]
+    pub fn load(&self, ordering: Ordering) -> Gc<T> {
+        let ptr;
+        let internal_handle;
+        {
+            let _collection_blocker = COLLECTOR.get_atomic_guard_spinlock_inclusive();
```

Comment:
Need to block collection here to prevent following race condition:

```
1) load reads handle (but doesn't create new handle yet)
2) write overrides handle, finishes operation
3) handle is garbage collected 
```
Now thread 1 has a dangling handle that it's going to turn into a Gc!